### PR TITLE
Remove `numpy` dependencies in `src/lightning/pytorch`

### PR DESCRIPTION
**Pull request by Peiffap**
## What does this PR do?
Removes dependency on `numpy` in part of the codebase (`src/lightning/pytorch`), as mentioned in #17278.

**Partially** fixes #17278

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19841.org.readthedocs.build/en/19841/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/strategies/launchers/multiprocessing.py [hunk 1]
```diff
@@ -253,7 +252,9 @@ def update_main_process_results(self, trainer: "pl.Trainer", extra: Dict[str, An
         """
         # NOTE: `get_extra_results` needs to be called before
         callback_metrics = extra["callback_metrics"]
-        trainer.callback_metrics.update(apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x)))
+        trainer.callback_metrics.update(
+            apply_to_collection(callback_metrics, type(Tensor().numpy()), lambda x: torch.tensor(x))
```
Comment (Peiffap) [inline src/lightning/pytorch/strategies/launchers/multiprocessing.py:256]: Is there a cleaner way to access numpy types (without `import numpy` obviously)?

Comment (awaelchli) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py:256]: I suggest to remove this conversion altogether (L255). The counterpart in L239 should then just clone the tensor instead of calling numpy(). This would then cover the case that is described in the comment and a numpy conversion is no longer needed.

Comment (Peiffap) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py]: @awaelchli I rebased and pushed a commit that does what I got from your comment. Let me know if this is what you meant.

Comment (awaelchli) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py]: Hey @Peiffap thank you for trying my suggestion. The multiprocessing tests don't pass though, so my suggestion seems wrong. I haven't looked at it closely, so I don't know what's going on or what the solution could be. It might be good to remove the code changes from `multiprocessing.py` here and merge the PR, leaving this for a future investigation.

Comment (Peiffap) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py]: Ah, I see. Thanks for giving it a shot at least. I've reverted the changes and I'll open an issue to see how to avoid casting to numpy without breaking things.

Comment (awaelchli) [reply src/lightning/pytorch/strategies/launchers/multiprocessing.py]: perfect!

**Comment by awaelchli** [comment]
Thanks a lot for your recent PRs @Peiffap!
